### PR TITLE
feature: hugepages support

### PR DESCRIPTION
**Pull request by liubin**
Porting of:

- kata-containers/agent#872
- kata-containers/runtime#3109

Fixes: #3342

### CODE REVIEW

File: src/agent/src/mount.rs [hunk 1]
```diff
@@ -289,6 +290,82 @@ async fn virtio9p_storage_handler(
...
+
+    file.write_all(numpages.as_bytes())
+        .context(format!("write nr_hugepages faile: {:?}", &path))?;
+
+    Ok(())
+}
+
+// Parse filesystem options string to retrieve hugepage details
+// options eg "pagesize=2048,size=107374182"
+fn get_pagesize_and_size_from_option(options: &Vec<String>) -> Result<(String, String)> {
```
Comment (jodh-intel) [inline src/agent/src/mount.rs:339]: Why not just return `Result<(usize, usize)>`?

File: src/agent/src/mount.rs [hunk 2]
```diff
@@ -289,6 +290,82 @@ async fn virtio9p_storage_handler(
...
+    let mut pagesize_str = "";
+    let mut size_str = "";
```
Comment (jodh-intel) [inline src/agent/src/mount.rs:341]: An `Option` would be cleaner here. Something like:

```rust
let mut pagesize: Option<&str> = None;

   :

pagesize = Some(stripped);
```

That way you don't need `found` either as you could say:

```rust
if pagesize.is_none() || size.is_none() {
   return Err(anyhow!("no pagesize/size options found"));
}
```

File: src/agent/src/mount.rs [hunk 3]
```diff
@@ -1380,4 +1460,56 @@ mod tests {
...
+
+    #[test]
+    fn test_get_pagesize_and_size_from_option() {
+        let expected_pagesize_str = "2048";
+        let expected_size_str = "107374182";
+        let expected = (expected_pagesize_str, expected_size_str);
+
+        let data = vec![
+            // (input, expected, is_ok)
+            ("size-1=107374182,pagesize-1=2048", expected, false),
```
Comment (jodh-intel) [inline src/agent/src/mount.rs:1472]: How about a few more tests:

- `size` but no `pagesize`.
- `pagesize` but no `size`.
- `pagesize` specified with invalid data.
- `size` specified with invalid data.
- Missing values (`pagesize=` and `size=`).

File: src/agent/src/mount.rs [hunk 4]
```diff
@@ -289,6 +290,82 @@ async fn virtio9p_storage_handler(
...
+    common_storage_handler(logger, storage)
+}
+
+// Allocate hugepages by writing to sysfs
+fn allocate_hugepages(logger: &Logger, options: &Vec<String>) -> Result<()> {
+    info!(logger, "mounting hugePages storage options: {:?}", options);
+
+    let (pagesize_str, size_str) = get_pagesize_and_size_from_option(options)?;
+    let pagesize = pagesize_str.parse::<i64>()?;
+    let size = size_str.parse::<i64>()?;
```
Comment (jodh-intel) [inline src/agent/src/mount.rs:314]: Shouldn't these be converting to `u64` or `usize`?

File: src/agent/src/mount.rs [hunk 5]
```diff
@@ -289,6 +290,82 @@ async fn virtio9p_storage_handler(
...
+    let numpages = format!("{}", size / pagesize);
+
+    let mut file = OpenOptions::new()
+        .create(true)
+        .write(true)
+        .truncate(true)
+        .open(&path)?;
+
+    file.write_all(numpages.as_bytes())
+        .context(format!("write nr_hugepages faile: {:?}", &path))?;
```
Comment (jodh-intel) [inline src/agent/src/mount.rs:332]: Nit: typo: "failed".

**Review by jodh-intel — COMMENTED** [review]
Thanks @liubin - a few questions.

File: src/agent/src/mount.rs (hunk 1 above)
Comment (liubin) [reply src/agent/src/mount.rs:339]: @jodh-intel this is just golang's translation, first I will let it work, and then make it more ideal.

**Comment by jodh-intel** [comment]
Ping @liubin.

File: src/agent/src/mount.rs [hunk 6]
```diff
@@ -289,6 +295,100 @@ async fn virtio9p_storage_handler(
...
+
+    file.write_all(numpages.as_bytes())
+        .context(format!("write nr_hugepages failed: {:?}", &path))?;
+
+    Ok(())
+}
+
+// Parse filesystem options string to retrieve hugepage details
+// options eg "pagesize=2048,size=107374182"
+fn get_pagesize_and_size_from_option(options: &[String]) -> Result<(i64, i64)> {
```
Comment (jodh-intel) [inline src/agent/src/mount.rs:356]: Surely the returned values cannot be negative so why not return `(u64, u64)`?

Comment (liubin) [reply src/agent/src/mount.rs]: @jodh-intel Changed to `u64`.

File: src/agent/src/mount.rs [hunk 7]
```diff
@@ -289,6 +295,100 @@ async fn virtio9p_storage_handler(
...
+    // /sys/kernel/mm/hugepages/hugepages-1048576kB/nr_hugepages
+    // /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages
+    // options eg "pagesize=2097152,size=524288000"(2M, 500M)
+    allocate_hugepages(logger, &storage.options.to_vec()).context("allocate hugepages")?;
+
+    common_storage_handler(logger, storage)?;
+
+    // hugetlbfs return empty string as ephemeral_storage_handler do.
+    // this is a sandbox level storage, but not a container-level mount.
+    Ok("".to_string())
```
[the remark this answers is not in the document]
Comment (liubin) [reply src/agent/src/mount.rs:311]: Other handlers are returning empty strings for special storage, fixing this should be another issue.

File: src/agent/src/mount.rs [hunk 8]
```diff
@@ -289,6 +295,100 @@ async fn virtio9p_storage_handler(
...
+        );
+    }
+
+    // write numpages to nr_hugepages file.
+    let path = path.join("nr_hugepages");
+    let numpages = format!("{}", size / pagesize);
+    info!(logger, "write {} pages to {:?}", &numpages, &path);
+
+    let mut file = OpenOptions::new()
+        // .create(true)  FIXME: does this file always exists?
```
[the remark this answers is not in the document]
Comment (liubin) [reply src/agent/src/mount.rs]: All FIXMEs have been fixed.

**Comment by fidencio** [comment]
A head's up that may affect your PR.

In the Architecture Committee meeting from January 25th, 2022, the Architecture Committee has agreed on using the "Dismiss stale pull request approvals when new commits are pushed" configuration from GitHub.  It basically means that if your PR has been rebased or updated, the approvals given will be erased.

In order to minimize traumas due to the new approach, please, consider adding a note on the changes done before the rebase / force-push, and also pinging the reviewers for a subsequent round of reviews.

Thanks for your understanding!

Related issue: https://github.com/kata-containers/community/issues/249

File: src/runtime/virtcontainers/kata_agent.go [hunk 9]
```diff
@@ -1358,6 +1370,13 @@ func (k *kataAgent) createContainer(ctx context.Context, sandbox *Sandbox, c *Co
 
 	ctrStorages = append(ctrStorages, epheStorages...)
 
+	k.Logger().WithField("ociSpec Hugepage Resources", ociSpec.Linux.Resources.HugepageLimits).Debug("ociSpec HugepageLimit")
+	hugepages, err := k.handleHugepages(ociSpec.Mounts, ociSpec.Linux.Resources.HugepageLimits)
```
Comment (bergwolf) [inline src/runtime/virtcontainers/kata_agent.go:1374]: Is `/sys/kernel/mm/hugepages/` per container? This seems to set a per container hugepage limit. What happens if two containers both set it?

**Review by bergwolf — DISMISSED** [review]

File: src/runtime/virtcontainers/kata_agent.go (hunk 9 above)
Comment (liubin) [reply src/runtime/virtcontainers/kata_agent.go:1374]: Now, this PR is a directly porting of the original PRs, modifying of `/sys/kernel/mm/hugepages/` seems like over-writing but not accumulating, so I think the second container's resource limitation will overwrite the first one.

Also both 2M and 1G in one Pod are not tested in these PRs(including PR in tests repo).

**Comment by jodh-intel** [comment]
Hi @liubin - please can you update to resolve the conflict?

**Comment by liubin** [comment]
/test
PATL @jodh-intel @bergwolf @lifupan

**Comment by liubin** [comment]
CI/GitHub actions errors is caused by GitHub issues:

> 14:20:16 fatal: unable to access 'https://github.com/tianocore/edk2.git/': Failed to connect to github.com port 443: Connection timed out

**Review by bergwolf — APPROVED** [review]

**Review by jodh-intel — APPROVED** [review]
Thanks @liubin.

lgtm

**Pull request merged**